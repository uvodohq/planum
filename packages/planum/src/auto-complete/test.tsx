import {
  autoUpdate,
  size,
  useDismiss,
  useFloating,
  useId,
  useInteractions,
  useListNavigation,
  useRole,
} from '@floating-ui/react-dom-interactions'
import React, { forwardRef, useLayoutEffect, useRef, useState } from 'react'

interface ItemProps {
  children: React.ReactNode
  active: boolean
}

const Item = forwardRef<
  HTMLDivElement,
  ItemProps & React.HTMLProps<HTMLDivElement>
>(({ children, active, ...rest }, ref) => {
  const id = useId()
  return (
    <div
      ref={ref}
      role="option"
      id={id}
      aria-selected={active}
      {...rest}
      style={{
        background: active ? 'lightblue' : 'none',
        padding: 4,
        cursor: 'default',
        ...rest.style,
      }}>
      {children}
    </div>
  )
})
Item.displayName = 'Item'

export function AutoCompleteTest({ data }) {
  const [open, setOpen] = useState(false)
  const [inputValue, setInputValue] = useState('')
  const [activeIndex, setActiveIndex] = useState<number | null>(null)

  const listRef = useRef<Array<HTMLElement | null>>([])

  const { x, y, reference, floating, strategy, context, refs } =
    useFloating<HTMLInputElement>({
      whileElementsMounted: autoUpdate,
      open,
      onOpenChange: setOpen,
      middleware: [
        size({
          apply({ rects, elements }) {
            Object.assign(elements.floating.style, {
              width: `${rects.reference.width}px`,
              maxHeight: `${200}px`,
            })
          },
          padding: 10,
        }),
      ],
    })

  useLayoutEffect(() => {
    // IMPORTANT: When the floating element first opens, this runs when the
    // styles have **not yet** been applied to the element. This can cause an
    // infinite loop as `size` has not yet limited the maxHeight, so the whole
    // page tries to scroll. We must wrap it in rAF.
    requestAnimationFrame(() => {
      if (activeIndex != null) {
        listRef.current[activeIndex]?.scrollIntoView({ block: 'nearest' })
      }
    })
  }, [activeIndex])

  const { getReferenceProps, getFloatingProps, getItemProps } = useInteractions(
    [
      useRole(context, { role: 'listbox' }),
      useDismiss(context),
      useListNavigation(context, {
        listRef,
        activeIndex,
        onNavigate: setActiveIndex,
        virtual: true,
        loop: true,
        focusItemOnHover: false,
      }),
    ],
  )

  function onChange(event: React.ChangeEvent<HTMLInputElement>) {
    const value = event.target.value
    setInputValue(value)

    if (value) {
      setOpen(true)
      setActiveIndex(0)
    } else {
      setOpen(false)
    }
  }

  const items = data.filter((item) =>
    item.toLowerCase().startsWith(inputValue.toLowerCase()),
  )

  return (
    <>
      <input
        {...getReferenceProps({
          ref: reference,
          onChange,
          value: inputValue,
          placeholder: 'Enter fruit',
          'aria-autocomplete': 'list',
          onKeyDown(event) {
            if (
              event.key === 'Enter' &&
              activeIndex != null &&
              items[activeIndex]
            ) {
              setInputValue(items[activeIndex])
              setActiveIndex(null)
              setOpen(false)
            }

            if (event.key === 'Tab') {
              setOpen(false)
            }
          },
        })}
      />
      {open && (
        <div
          {...getFloatingProps({
            ref: floating,
            style: {
              position: strategy,
              left: x ?? 0,
              top: y ?? 0,
              background: '#eee',
              color: 'black',
              overflowY: 'auto',
            },
          })}>
          {items.map((item, index) => (
            // eslint-disable-next-line react/jsx-key
            <Item
              {...getItemProps({
                key: item,
                ref(node) {
                  listRef.current[index] = node
                },
                onClick() {
                  setInputValue(item)
                  setOpen(false)
                  refs.reference.current?.focus()
                },
              })}
              active={activeIndex === index}>
              {item}
            </Item>
          ))}
        </div>
      )}
    </>
  )
}
