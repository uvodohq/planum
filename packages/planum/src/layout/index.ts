export type { SpacerProps, StackProps } from './layout'
export { Box, Flex, Spacer, Stack } from './layout'
