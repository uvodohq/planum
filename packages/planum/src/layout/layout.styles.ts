import { styled } from '../theme'

export const StyledStack = styled('div', {
  flex: 1,
  justifySelf: 'stretch',
  alignSelf: 'stretch',
})
