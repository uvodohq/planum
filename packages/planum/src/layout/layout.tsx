import * as React from 'react'

import type { CSS, SpacingValue } from '../theme'
import { styled } from '../theme'

export const Box = styled('div', {})

export type StackProps = React.ComponentProps<typeof Box> & {
  y?: SpacingValue
  x?: SpacingValue
}

export const Stack = (props: StackProps) => {
  const { x, y, ...rest } = props

  return (
    <Box {...rest}>
      {React.Children.map(props.children, (child) => {
        return (
          <Box
            css={{
              '&:not(:last-child)': {
                mb: y,
                mr: x,
                margin: '',
              },
            }}>
            {child}
          </Box>
        )
      })}
    </Box>
  )
}

interface FlexProps {
  dir?: 'row' | 'column'
  justify?: 'start' | 'end' | 'center' | 'space-between' | 'space-around'
  align?: 'start' | 'end' | 'center' | 'stretch' | 'baseline'
  wrap?: 'nowrap' | 'wrap' | 'wrap-reverse'
  center?: boolean
  gap?: SpacingValue
  css?: CSS
  as?: React.ElementType
  [key: string]: any
}

export const Flex = (props: FlexProps) => {
  const { center, gap, align, justify, wrap, css = {}, children } = props

  return (
    <Box
      css={{
        display: 'flex',
        gap,
        alignItems: align ?? center ? 'center' : '',
        justifyContent: justify ?? center ? 'center' : '',
        flexWrap: wrap,

        ...(css as any),
      }}>
      {children}
    </Box>
  )
}

export interface SpacerProps extends StackProps {}

export const Spacer = (props: SpacerProps) => {
  const { x, y, css = {}, ...rest } = props

  return (
    <Box
      css={{
        height: y,
        width: x,
        minWidth: x,
        ...(css as any),
      }}
      {...rest}
    />
  )
}
