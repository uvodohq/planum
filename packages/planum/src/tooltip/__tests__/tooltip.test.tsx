import { Box } from '../../layout'
import { render, screen, userEvent, waitFor } from '../../test'
import { Tooltip, TooltipPopup, TooltipTrigger } from '..'
// import type { UseTooltipProps } from '../use-tooltip'

global.ResizeObserver = require('resize-observer-polyfill')

function renderTooltip(props: Omit<any, 'children'>) {
  return render(
    <Tooltip {...props}>
      <TooltipTrigger>
        <Box css={{ bg: 'red' }} data-testid="trigger">
          trigger
        </Box>
      </TooltipTrigger>
      <TooltipPopup>text</TooltipPopup>
    </Tooltip>,
  )
}

const label = 'tooltip'

describe('<Tooltip />', async () => {
  it('should render the tooltip trigger without popup content', async () => {
    renderTooltip({ label })

    const tooltipTrigger = screen.getByTestId('trigger')

    expect(tooltipTrigger).toBeInTheDocument()
    // get snapshot of the tooltip trigger
    expect(tooltipTrigger).toMatchSnapshot()
  })

  it('should show tooltip when hover over trigger', async () => {
    renderTooltip({ label })

    const tooltipTrigger = screen.getByTestId('trigger')

    await userEvent.hover(tooltipTrigger)

    await waitFor(() => {
      expect(screen.getByRole('tooltip')).toBeInTheDocument()
    })
  })
})
