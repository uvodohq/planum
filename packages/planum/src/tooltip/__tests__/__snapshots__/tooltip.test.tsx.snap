// Vitest Snapshot v1

exports[`<Tooltip /> > should render the tooltip trigger without popup content 1`] = `
<div
  class="Planum-c-PJLV Planum-c-PJLV-ibzwKCF-css"
  data-testid="trigger"
>
  trigger
</div>
`;
